internal/resource/buffer.go

```diff
@@ -0,0 +1,120 @@
+// Copyright 2017 CoreOS, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package resource
+
+import (
+	"errors"
+	"io"
+	"sync"
+)
+
+var (
+	ErrOutOfMemory    = errors.New("unable to grow buffer")
+	ErrOffsetTooLarge = errors.New("cannot seek past end of buffer")
+	ErrUnknownWhence  = errors.New("unknown whence")
+)
+
+type buffer struct {
+	buf         []byte
+	cursor      int64
+	end         int64
+	cap         int64
+	writeLock   sync.Mutex
+	writeAtLock sync.Mutex
+}
+
+func NewBuffer() *buffer {
+	return &buffer{}
+}
+
+func (b *buffer) grow(n int64) error {
+	newCap := 2*b.cap + n
+	buf := make([]byte, newCap)
+	copy(buf, b.buf)
+	b.buf = buf
+	b.cap = newCap
+	return nil
+}
+
+func (b *buffer) Write(p []byte) (int, error) {
+	b.writeLock.Lock()
+	defer b.writeLock.Unlock()
+
+	n, err := b.WriteAt(p, b.cursor)
+	if err != nil {
+		return 0, err
+	}
+	b.cursor += int64(len(p))
+	if b.end < b.cursor {
+		b.end = b.cursor
+	}
+	return n, nil
+}
+
+func (b *buffer) WriteAt(p []byte, off int64) (int, error) {
+	b.writeAtLock.Lock()
+	defer b.writeAtLock.Unlock()
+
+	lastWritePos := off + int64(len(p))
+	if lastWritePos > int64(b.cap) {
+		err := b.grow(lastWritePos)
+		if err != nil {
+			return 0, err
+		}
+	}
+	if lastWritePos > b.end {
+		b.end = lastWritePos
+	}
+	copy(b.buf[off:], p)
+	return len(p), nil
+}
+
+func (b *buffer) Seek(offset int64, whence int) (int64, error) {
+	switch whence {
+	case io.SeekStart:
+		if offset > b.end {
+			return b.cursor, ErrOffsetTooLarge
+		}
+		b.cursor = offset
+		return b.cursor, nil
+	case io.SeekCurrent:
+		if b.cursor+offset > b.end {
+			return b.cursor, ErrOffsetTooLarge
+		}
+		b.cursor += offset
+		return b.cursor, nil
+	case io.SeekEnd:
+		if offset > 0 {
+			return b.cursor, ErrOffsetTooLarge
+		}
+		b.cursor += offset
+		return b.cursor, nil
+	default:
+		return b.cursor, ErrUnknownWhence
+	}
+}
+
+func (b *buffer) Read(p []byte) (int, error) {
+	if b.cursor == b.end {
+		return 0, io.EOF
+	}
+	n := copy(p, b.buf[b.cursor:])
+	b.cursor += int64(n)
+	return n, nil
+}
+
+func (b *buffer) Bytes() []byte {
```

Comment:
I just tweaked the implementation to rename the existing mutexes and use them for reading and seeking operations.

When writing this I was only considering the ways in which it would be used by Ignition, where no reading/seeking is done until all writing is complete, and it would definitely be much better to have a more robust implementation.